dockerfiles/crystal-1.12.Dockerfile

```diff
@@ -0,0 +1,14 @@
+FROM crystallang/crystal:1.12-alpine
+
+ENV CODECRAFTERS_DEPENDENCY_FILE_PATHS="shard.yml, shard.lock"
+
+WORKDIR /app
+COPY shard.yml shard.lock ./
+
+RUN shards install --frozen
+
+RUN mkdir -p /app-cached/
+RUN if [ -d "/app/lib" ]; then mv /app/lib /app-cached/lib; fi
+
+RUN echo "cd \${CODECRAFTERS_SUBMISSION_DIR} && crystal build -p -s -t -o server app/main.cr && sed -i '/^crystal/ s/^/# /' ./spawn_redis_server.sh" > /codecrafters-precompile.sh
```

Comment:
Not pertinent to this PR, but just wanted to note: looks like this pattern is becoming commonplace in a lot of our languages. Need to find a better solution. This commenting stuff feels wonky - if a user edits the file it isn't going to be clear to them that there's a hidden script that relies on the file's structure.